### PR TITLE
ENH: add scipy.linalg.convolution_matrix

### PR DESCRIPTION
**Pull request by mborgerding**
Reference issue
See #7834 

#### What does this implement/fix?
This implements the `convolution_matrix` function. It is similar to the Matlab/Octave `convmtx` function but not derived from their implementations in any way.
There is an additional "mode" argument analogous to that in numpy.convolve.

### CODE REVIEW

File: scipy/linalg/special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/special_matrices.py]: I wonder if this is maybe a little challenging to read because i.e., `v` is defined in another docstring in NumPy. Maybe that's obvious to experts though.

File: scipy/linalg/tests/test_special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/tests/test_special_matrices.py]: This may be done more cleanly with a parametrized test, although it is true that we use this paradigm for testing elsewhere in a few cases.

File: scipy/linalg/special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/special_matrices.py]: the line coverage looks pretty good for the new function, but I did notice that Azure reports no test hitting this exception handling

File: scipy/linalg/special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/special_matrices.py]: If you search for the docstring render produced in CI ("build_docs artifact") and navigate to this new function, I think you'll notice that the formatting of these values is not quite right--they get kind of crammed on the same line

**Review by tylerjereddy — COMMENTED** [review]
Added a few comments.

Note that we normally discuss new features on the mailing list, but I see that this actually already has some positive feedback in the cross-linked issue. The longer-term issue was finding an implementation that was not in another license-restricted code base.

Is there a citation for this implementation?

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: Good point.  I'll add the mathematical definition.

File: scipy/linalg/tests/test_special_matrices.py
Comment (mborgerding) [reply scipy/linalg/tests/test_special_matrices.py]: In order to get the same outer product coverage it now has, would one nest the `pytest.mark.parameterize` decorators?

Can you provide a reference or example?

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: Ewww! .. Yeah I'll fix that.  

I'll push later tonight or tomorrow addressing your above excellent comments.

File: scipy/linalg/special_matrices.py
Comment (ilayn) [inline scipy/linalg/special_matrices.py]: Here we need a validation of the mode argument via say
```
    if mode not in ['full', 'same', 'valid']:
        raise ValueError("'mode' argument must be one of the 'full', 'same',"
                         " 'valid'")
```
Then we can safely use the `else` block below.

File: scipy/linalg/tests/test_special_matrices.py
Comment (ilayn) [inline scipy/linalg/tests/test_special_matrices.py]: Why do you need specifically gaussian distribution here? Any random vector should work here right? Also test files also better have PEP8 compatibility.

File: scipy/linalg/tests/test_special_matrices.py
Comment (ilayn) [reply scipy/linalg/tests/test_special_matrices.py]: Indeed nesting is the typical way. 

```
@pytest.mark.parameterize('cpx', (False, True))
@pytest.mark.parameterize('na', range(2, 7))
@pytest.mark.parameterize('nv', range(2, 7))
@pytest.mark.parameterize('mode', (None, 'full', 'valid', 'same'))
def test_convmtx(cpx, na, nv, mode):
    ....
    Test contents
    .....
```

However note that this product introduces 200 tests. Hence if it takes too long we might need to pick 10, 100, 1000 as sample sizes and reduce the number of tests since they essentially test the same thing with minute differences. Testing time is unfortunately a constraint for us.

**Review by ilayn — CHANGES_REQUESTED** [review]
@mborgerding Thank you for your contribution and time. I've added some comments which all are of minor nature. My only major comment is the function name that it is probably better to spell out instead of the typical matlab cryptic naming scheme. I propose `convolution_matrix` but maybe you or others have a better name proposal.

File: scipy/linalg/tests/test_special_matrices.py
Comment (mborgerding) [reply scipy/linalg/tests/test_special_matrices.py]: In light of the increased number of tests, perhaps such refactoring could be postponed until there is a need for more granular tests?

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: I've put both `A @ v` and `matmul(A,v)` into the docstring.  The former is more modern, the latter is more descriptive and easier to research.

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: done

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: done

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: done

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: done

File: scipy/linalg/tests/test_special_matrices.py
Comment (mborgerding) [reply scipy/linalg/tests/test_special_matrices.py]: Does not need to be Gaussian, per se.  I've updated the names to clarify this.  I added a double line above the function def.  If you are referring to something else that is not PEP8, I'm missing it (as does `autopep8`)

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: mode restricted as of 2a1bd974e5605d103028f8b64919d30ba900aa7d

File: scipy/linalg/tests/test_special_matrices.py
Comment (ilayn) [reply scipy/linalg/tests/test_special_matrices.py]: I meant that the line is longer than 79 chars.

Comment (mborgerding) [reply scipy/linalg/tests/test_special_matrices.py]: fixed

**Comment by mborgerding** [comment]
@ilayn  , @tylerjereddy 
I believe all changes have been done except we have not resolved the discussion above about using `itertools.product` in a single fast test vs. nesting `parameterize` decorators to make multiple tests.  
I tend to favor the itertools approach as I  find it clearer to read and maintain the code, but I don't feel strongly about it.  I defer to your judgement.

**Comment by mborgerding** [comment]
bump

**Comment by ilayn** [comment]
Hi @mborgerding instead of range(2,7) for both test cases please use [10, 100] such that at least two order of magnitudes tested. There is not much difference between 3 and 4 in terms of testing behavior. Some problems only reveal themselves when the values are large or arrays size are large.

**Comment by mborgerding** [comment]
> Hi @mborgerding instead of range(2,7) for both test cases please use [10, 100] such that at least two order of magnitudes tested. There is not much difference between 3 and 4 in terms of testing behavior. Some problems only reveal themselves when the values are large or arrays size are large.

I expanded the range, but rather than [10,100],  it tests the lowest meaningful number for convolution (2) and at least one odd number (9), and 100.

File: scipy/linalg/special_matrices.py
Comment (ilayn) [inline scipy/linalg/special_matrices.py]: PEP8: whitespace after every comma

File: scipy/linalg/special_matrices.py
Comment (ilayn) [inline scipy/linalg/special_matrices.py]: PEP8: whitespace after every comma

File: scipy/linalg/special_matrices.py
Comment (ilayn) [inline scipy/linalg/special_matrices.py]: PEP8: whitespace after every comma

File: scipy/linalg/special_matrices.py
Comment (ilayn) [inline scipy/linalg/special_matrices.py]: PEP8: line too long

File: scipy/linalg/tests/test_special_matrices.py
Comment (ilayn) [inline scipy/linalg/tests/test_special_matrices.py]: PEP8: line too long

File: scipy/linalg/special_matrices.py
Comment (ilayn) [inline scipy/linalg/special_matrices.py]: All codelike backtick guards need to be doubled as single backtick means a Sphinx link directive to link to other functions for example to have a link to another function \`scipy.linalg.toeplitz\`

**Review by ilayn — CHANGES_REQUESTED** [review]
You can install pycodestyle and use that for trivial PEP8 checks or adjust your editor such that it does it for you.

**Comment by mborgerding** [comment]
> You can install pycodestyle and use that for trivial PEP8 checks or adjust your editor such that it does it for you.

I appreciate the tip, but the default pycodestyle options did not seem to catch some of the commas.
Apparently some of the checks are turned off by default.
I was able to turn on full pedantic mode with "--ignore ''"

File: scipy/linalg/special_matrices.py
Comment (mborgerding) [reply scipy/linalg/special_matrices.py]: OK, I think I've followed the single-vs-double backtick convention

**Comment by mborgerding** [comment]
Is there anything more to be done for this PR?

File: scipy/linalg/special_matrices.py
Comment (ilayn) [inline scipy/linalg/special_matrices.py]: This one looks like the only offending PEP8 item.

**Review by ilayn — CHANGES_REQUESTED** [review]

**Comment by ilayn** [comment]
If the CI is green this would be ready for 1.4.0. Please review if you would have time. I'll add the label prospectively.

**Comment by ilayn** [comment]
@mborgerding 

Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches.

**Review by ilayn — APPROVED** [review]

File: scipy/linalg/tests/test_special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/tests/test_special_matrices.py]: we try to avoid these slash continuations usually -- I guess that can be cleaned up later

File: scipy/linalg/tests/test_special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/tests/test_special_matrices.py]: normally we'd use a `class` and then place specific tests within that class for pytest namespace resolution

File: scipy/linalg/tests/test_special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/tests/test_special_matrices.py]: I suppose this is still used in our codebase a fair bit; at some number of permutations I think we should prefer to parametrize, but anyway..

**Review by tylerjereddy — COMMENTED** [review]
Was going to wait for Azure CI to finish to look for coverage holes, but if @ilayn is happy that should be good enough I suppose.

**Comment by tylerjereddy** [comment]
@rgommers Have you observed this Python 3.8/Windows Azure timeout issue before?

**Comment by tylerjereddy** [comment]
The `ppc64le` failure probably isn't related

**Comment by tylerjereddy** [comment]
There are no holes in the line coverage, which is good.

**Comment by rgommers** [comment]
> @rgommers Have you observed this Python 3.8/Windows Azure timeout issue before?

no I haven't

**Comment by tylerjereddy** [comment]
I'm bumping the milestone--we've already had more last-minute merges than I'd like & the CI is a little strange here, but this looks close to merging after we branch.

**Comment by ilayn** [comment]
The failures are not related to this PR.

File: scipy/linalg/tests/test_special_matrices.py
Comment (ilayn) [reply scipy/linalg/tests/test_special_matrices.py]: It's a singleton so here a class is not needed

Comment (tylerjereddy) [reply scipy/linalg/tests/test_special_matrices.py]: I don't understand--there are actually several things being tested below--normally one would split each thing being tested into its own specific **unit** test classified within the same test class.

File: scipy/linalg/tests/test_special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/tests/test_special_matrices.py]: This is one test

File: scipy/linalg/tests/test_special_matrices.py
Comment (tylerjereddy) [inline scipy/linalg/tests/test_special_matrices.py]: This is testing something else

Comment (ilayn) [reply scipy/linalg/tests/test_special_matrices.py]: That's unnecessary granularity. And also if there is no scope nesting it's just a preference. If we write one test per assertion we will have millions of tests.

Comment (ilayn) [reply scipy/linalg/tests/test_special_matrices.py]: See examples in the rest of the file for more examples.

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
@mborgerding, thanks for contributing this pull request.  `convolution_matrix` will be a nice addtion to the collection of special matrix functions, and I'd like to get it into version 1.5.0.

I made a bunch of comments in-line.  Here are some additional comments:

* Thanks for completing the table of special matrices in the tutorial!
* There are some incorrect results.
  In the following two examples, the output should be `3*np.eye(5)`:

      In [114]: convolution_matrix([3], 5, mode='valid')

      Out[114]: array([], shape=(0, 0), dtype=int64)

      In [115]: convolution_matrix([3], 5, mode='same')

      Out[115]: array([], shape=(0, 5), dtype=int64)

  In the next two examples, the output should be `np.array([[3], [4]])`:

      In [146]: convolution_matrix([3, 4], 1, mode='valid')

      Out[146]: array([], shape=(0, 0), dtype=int64)

      In [147]: convolution_matrix([3, 4], 1, mode='same')

      Out[147]: array([], shape=(0, 1), dtype=int64)

It has been about 5 months since there was activity in this pull request.  Do you still have time and interest for working on it?  Let us know.  If not, we can take it over.

File: scipy/linalg/tests/test_special_matrices.py
Comment (WarrenWeckesser) [reply scipy/linalg/tests/test_special_matrices.py:647]: I agree with @tylerjereddy, the tests should be split up, and the big loop at the end should use pytest's `parametrize` feature.  The increased granularity lets us know all the things that are going wrong if there is a failure, instead of just the first thing that goes wrong.

File: scipy/linalg/tests/test_special_matrices.py
Comment (WarrenWeckesser) [reply scipy/linalg/tests/test_special_matrices.py:663]: Unfortunately, there is a lot of code in our codebase that does not conform to what we now consider best practices.  It is our job as reviewers to ensure the *new* code does so.

**Comment by tylerjereddy** [comment]
> Do you still have time and interest for working on it? Let us know. If not, we can take it over.

@WarrenWeckesser Are you going to have time to take over here before we branch?

**Comment by WarrenWeckesser** [comment]
@tylerjereddy, thanks for the reminder.  Yes, I'll take this over.

**Comment by WarrenWeckesser** [comment]
Update pull request is in https://github.com/scipy/scipy/pull/12135.

**Comment by WarrenWeckesser** [comment]
This PR was continued in gh-12135, and that PR has been merged, so closing.

Thanks @mborgerding for the contribution, and thanks @ilayn and @tylerjereddy for the reviews.

**Pull request closed without merging**
